api/src/main/java/org/pmiops/workbench/db/dao/RdrExportDao.java

```diff
@@ -6,15 +6,20 @@
 import org.pmiops.workbench.db.model.DbRdrExport;
 import org.springframework.data.jpa.repository.Query;
 import org.springframework.data.repository.CrudRepository;
+import org.springframework.data.repository.query.Param;
 
 public interface RdrExportDao extends CrudRepository<DbRdrExport, Long> {
 
-  @Query(
-      nativeQuery = true,
-      value =
-          "select u.user_id from user u LEFT JOIN rdr_export rdr on"
-              + " u.user_id = rdr.export_id and rdr.entity_type = 1 where "
-              + "u.last_modified_time > rdr.last_Export_date or rdr.export_id is null")
+  public final String workspaceIdsToExportQuery =
+      "select w.workspace_id from  workspace w LEFT JOIN "
+          + "rdr_export rdr on w.workspace_id = rdr.entity_id and rdr.entity_type = 2 where "
+          + "w.last_modified_time > rdr.last_export_date or rdr.entity_id IS NULL";
+  public final String unchangedWorkspaceIdsQuery =
+      "select workspace_id from workspace where workspace_id not in ("
+          + workspaceIdsToExportQuery
```

Comment:
Let me restate my concern because I don't think it's been understood. I am worried that writing this query like `"IN (<arbitrary large list>)"` might simply fail if you try to run this when the inner list is large (when I said "memory" in my comment, I was not talking about memory on our API server - rather memory used on the SQL server). I'm not concerned about the number of elements returned overall from the query, or the performance of fetching the list in the subquery. In general, I don't care much about this being fast for this use case, really as long as it finishes it's fine.

I commented because it struck me as an unnatural way to write the query in general. `IN (...)` is generally used for small / bounded lists of elements, which is not necessarily the case here. I'm guessing this can probably work, but has to be done very carefully or else you have very degenerate situations where the inner query gets run once **per row** (you can google for correlated vs non-correlated subqueries).

I would just write this in a more standard way to start, so we don't run into issues later. If you want to maintain composability, I think it would be better to do one of the following:
1. Parameterize the query or pull out SQL strings to allow you to negate the `WHERE` clause
2. Have another method to get all ids, and do the Set subtraction in java.

Reply:
Thanks for the clarification. This is a non-correlated query since the subquery does not depend on its outer query. I will leave as-is, as the intent of the code seems clear with this approach.